### PR TITLE
DPLASMA Warmup -- 2nd try

### PR DESCRIPTION
**Pull request by therault**
This is a second try for solving the warmup issue in DPLASMA (especially in CUDA codes).

Here are some performance measurements of the approach proposed in this PR, on Leconte (8x V100):

dpotrf-leconte-warmup-vs-nowarmup-avg-white

dpotrf-leconte-warmup-vs-nowarmup-details-white

'gflops/avg' represents the ratio 'gflops of this run' divided by the 'appropriate average' (meaning the average without the outlier on runs without warmup, and the average of all measured points on runs with warmup).

There is still some warmup problem that is unidentified, at large tile size (512, 1024), for 1 to 4 GPUs, the first actual run is still slower than the others for the small problem sizes. It's unclear what is the source of the issue at this point, but the warmup patch fixes most of the CUDA/CUBLAS warmup issues.

The goal of the current code is to include changes for all tests that feature a CUDA implementation and timing:
  -  POTRF
  -  GEMM (WIP)
  -  POINV
  -  GEQRF

TRSM is the last kernel that features a CUDA implementation, and it does not have timing in its testings.

Aurelien has notified during the discusssion of an issue with HIP: allocation of memory on the HIP device was lazy at some point, and allocation at first touch is a significant part of the warmup overheads of the HIP runs. We decided that this should be solved at the PaRSEC level, during memory allocation, and not at the DPLASMA warmup level.

### CODE REVIEW

**Comment by therault** [comment]
Done: all PTG tests that are CUDA-enabled

To do:
 - DTD tests that are CUDA-enabled
 - Update performance of CUDA-enabled tests on recent machines

**Comment by bosilca** [comment]
As discussed on 03/31/23 we need to

- [ ]  cover the CPU only tests (in addition to the device tests)

- [ ] need to touch the entire data at least once on the GPU (at the level of PaRSEC data during the allocation stage).

**Comment by therault** [comment]
- [] check what is happening in paranoid debug mode: tests on frontier show that the local data dists generated via the warmup calls trigger (wrongful?) asserts
- [] GPU load statistics should be reset after the completion of each warmup test. Shouldn't that happen automatically via the completion of tasks?

File: src/zgeqrf.jdf [hunk 1]
```diff
@@ -495,6 +495,8 @@ BODY [type=CUDA device=%{ return n; %}
                          WORK,  ib,
                          WORKC, descA->mb,
                          parsec_body.stream );
+
```
Comment (abouteiller) [inline src/zgeqrf.jdf:498]: sure?

**Review by abouteiller — APPROVED** [review]

File: src/zgeqrf.jdf [hunk 2]
```diff
@@ -495,6 +495,8 @@ BODY [type=CUDA device=%{ return n; %}
...
+      parsec_gpu_push_workspace(gpu_device, gpu_stream);
```
Comment (abouteiller) [inline src/zgeqrf.jdf:499]: sure?

**Review by abouteiller — APPROVED** [review]

File: src/zgeqrf.jdf (hunk 2 above)
Comment (therault) [reply src/zgeqrf.jdf:499]: It's not directly related to the warmup fix, but I discovered the bug while testing warmup, and it's a 1-line fix, so I left it here.

It is required to push_workspace after the pop_workspace within the task, or we reach workspace exhaustion and assert or segfault.

**Pull request merged**